This is a continuation of the project to process covid data and see whether we can learn more about what causes covid deaths. The dataset for covid was from our world in data. It has been cleaned, explored, and most of the preprocessing has been done including scaling it.

We shall now import the data into the notebook:

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression 
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
df = pd.read_csv('../../Unit16FeatureEngineering/Capstone2Preprocessing/ScaledData.csv')

In [3]:
df.head()

date  new_cases  new_deaths  new_cases_per_million  \
0  2020-01-22  -0.684809   -1.230281              -0.684809   
1  2020-01-23  -0.684809   -1.230281              -0.684809   
2  2020-01-24  -0.684802   -1.230281              -0.684802   
3  2020-01-25  -0.684809   -1.230281              -0.684809   
4  2020-01-26  -0.684788   -1.230281              -0.684788   

   reproduction_rate  icu_patients  icu_patients_per_million  hosp_patients  \
0          -2.064156     -1.252235                 -1.252234      -1.139472   
1          -2.064156     -1.252235                 -1.252234      -1.139472   
2          -2.064156     -1.252235                 -1.252234      -1.139472   
3          -2.064156     -1.252235                 -1.252234      -1.139472   
4          -2.064156     -1.252235                 -1.252234      -1.139472   

   hosp_patients_per_million  weekly_hosp_admissions  ...  \
0                  -1.139472               -1.121894  ...   
1                  -1.139472               -1.121894  ...   
2                  -1.139472               -1.121894  ...   
3                  -1.139472               -1.121894  ...   
4                  -1.139472               -1.121894  ...   

   people_fully_vaccinated  total_boosters  new_vaccinations  \
0                -0.940899       -0.500739         -0.758638   
1                -0.940899       -0.500739         -0.758638   
2                -0.940899       -0.500739         -0.758638   
3                -0.940899       -0.500739         -0.758638   
4                -0.940899       -0.500739         -0.758638   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                       -0.936484                      -0.983961   
1                       -0.936484                      -0.983961   
2                       -0.936484                      -0.983961   
3                       -0.936484                      -0.983961   
4                       -0.936484                      -0.983961   

   people_fully_vaccinated_per_hundred  total_boosters_per_hundred  \
0                              -0.9409                   -0.497097   
1                              -0.9409                   -0.497097   
2                              -0.9409                   -0.497097   
3                              -0.9409                   -0.497097   
4                              -0.9409                   -0.497097   

   stringency_index  tests_units_not performed  tests_units_tests performed  
0         -3.781973                   4.460654                    -4.460654  
1         -3.781973                   4.460654                    -4.460654  
2         -3.781973                   4.460654                    -4.460654  
3         -3.781973                   4.460654                    -4.460654  
4         -3.781973                   4.460654                    -4.460654  

[5 rows x 29 columns]

In [4]:
df['date'] = pd.to_datetime(df['date'])

In [5]:
df.set_index('date', inplace=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 815 entries, 2020-01-22 to 2022-04-15
Data columns (total 28 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   new_cases                            815 non-null    float64
 1   new_deaths                           815 non-null    float64
 2   new_cases_per_million                815 non-null    float64
 3   reproduction_rate                    815 non-null    float64
 4   icu_patients                         815 non-null    float64
 5   icu_patients_per_million             815 non-null    float64
 6   hosp_patients                        815 non-null    float64
 7   hosp_patients_per_million            815 non-null    float64
 8   weekly_hosp_admissions               815 non-null    float64
 9   weekly_hosp_admissions_per_million   815 non-null    float64
 10  total_tests                          815 non-null    float64
 11  new_tests    

In [9]:
X = df.drop(['new_deaths'], axis=1)
y = df[['new_deaths']]

In [10]:
X

new_cases  new_cases_per_million  reproduction_rate  icu_patients  \
date                                                                            
2020-01-22  -0.684809              -0.684809          -2.064156     -1.252235   
2020-01-23  -0.684809              -0.684809          -2.064156     -1.252235   
2020-01-24  -0.684802              -0.684802          -2.064156     -1.252235   
2020-01-25  -0.684809              -0.684809          -2.064156     -1.252235   
2020-01-26  -0.684788              -0.684788          -2.064156     -1.252235   
...               ...                    ...                ...           ...   
2022-04-11  -0.383126              -0.383126           0.797451     -1.072014   
2022-04-12  -0.491969              -0.491969           0.797451     -1.076960   
2022-04-13  -0.399617              -0.399618           0.797451     -1.077305   
2022-04-14  -0.307182              -0.307181           0.797451     -1.074660   
2022-04-15  -0.413762              -0.413761           0.797451     -1.252235   

            icu_patients_per_million  hosp_patients  \
date                                                  
2020-01-22                 -1.252234      -1.139472   
2020-01-23                 -1.252234      -1.139472   
2020-01-24                 -1.252234      -1.139472   
2020-01-25                 -1.252234      -1.139472   
2020-01-26                 -1.252234      -1.139472   
...                              ...            ...   
2022-04-11                 -1.072011      -0.881618   
2022-04-12                 -1.076950      -0.881900   
2022-04-13                 -1.077294      -0.881234   
2022-04-14                 -1.074652      -0.879108   
2022-04-15                 -1.252234      -1.139472   

            hosp_patients_per_million  weekly_hosp_admissions  \
date                                                            
2020-01-22                  -1.139472               -1.121894   
2020-01-23                  -1.139472               -1.121894   
2020-01-24                  -1.139472               -1.121894   
2020-01-25                  -1.139472               -1.121894   
2020-01-26                  -1.139472               -1.121894   
...                               ...                     ...   
2022-04-11                  -0.881618               -0.835729   
2022-04-12                  -0.881899               -0.832645   
2022-04-13                  -0.881234               -0.830517   
2022-04-14                  -0.879111               -0.828634   
2022-04-15                  -1.139472               -1.121894   

            weekly_hosp_admissions_per_million  total_tests  ...  \
date                                                         ...   
2020-01-22                           -1.121894    -1.254081  ...   
2020-01-23                           -1.121894    -1.254081  ...   
2020-01-24                           -1.121894    -1.254081  ...   
2020-01-25                           -1.121894    -1.254081  ...   
2020-01-26                           -1.121894    -1.254081  ...   
...                                        ...          ...  ...   
2022-04-11                           -0.835724    -1.254081  ...   
2022-04-12                           -0.832645    -1.254081  ...   
2022-04-13                           -0.830519    -1.254081  ...   
2022-04-14                           -0.828629    -1.254081  ...   
2022-04-15                           -1.121894    -1.254081  ...   

            people_fully_vaccinated  total_boosters  new_vaccinations  \
date                                                                    
2020-01-22                -0.940899       -0.500739         -0.758638   
2020-01-23                -0.940899       -0.500739         -0.758638   
2020-01-24                -0.940899       -0.500739         -0.758638   
2020-01-25                -0.940899       -0.500739         -0.758638   
2020-01-26                -0.940899       -0.500739         -0.758638   


In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=40)

In [12]:
lm = LinearRegression()
lm.fit(X_train, y_train)
y_pred = lm.predict(X_test)

In [13]:
print(mean_squared_error(y_test, y_pred))
(r2_score(y_test, y_pred))

0.3765299802356845


0.6522232636314598

In [14]:
lm.get_params()

{'copy_X': True,
 'fit_intercept': True,
 'n_jobs': None,
 'normalize': False,
 'positive': False}

In [15]:
lr = LinearRegression()
grid_search = GridSearchCV(lr, param_grid={'copy_X': [True,False], 'fit_intercept': [True, False], 'n_jobs': [-1], 'normalize': [False, True], 'positive': [False, True]}, cv=5)

In [16]:
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=LinearRegression(),
             param_grid={'copy_X': [True, False],
                         'fit_intercept': [True, False], 'n_jobs': [-1],
                         'normalize': [False, True],
                         'positive': [False, True]})

In [17]:
y_pred2 = grid_search.predict(X_test)

In [18]:
print(mean_squared_error(y_test, y_pred2))
(r2_score(y_test, y_pred2))

0.3754878693165293


0.653185795045747

The r2 score of the model is at 65.3%. This is not a very high accuracy, but it is better than the mean.

As mentioned in the EDA notebook, the most telling features from this dataset are total_tests_per_thousand, people_fully_vaccinated, tests_units_not performed, positive_rate, reproduction_rate, new_vaccinations.